common/src/main/java/com/wynntils/models/stats/StatCalculator.java

```diff
@@ -0,0 +1,175 @@
+            }
+
+            int lowerRollBound = (int) Math.max(Math.ceil(lowerRawRollBound), starMin);
+            int higherRollBound = (int) Math.min(Math.ceil(higherRawRollBound) - 1, starMax);
+            return RangedValue.of(lowerRollBound, higherRollBound);
+        } else {
+            int lowerRollBound = (int) Math.min(Math.ceil(lowerRawRollBound) - 1, 130);
+            int higherRollBound = (int) Math.max(Math.ceil(higherRawRollBound), 70);
+            return RangedValue.of(lowerRollBound, higherRollBound);
+        }
+    }
+
+    public static int calculateStarsFromInternalRoll(int internalRoll) {
+        // Star calculation reference, from salted:
+        // https://forums.wynncraft.com/threads/about-the-little-asterisks.147931/#post-1654183
+        int stars;
+        if (internalRoll < 101) {
+            stars = 0;
+        } else if (internalRoll < 125) {
+            stars = 1;
+        } else if (internalRoll < 130) {
+            stars = 2;
+        } else {
+            stars = 3;
+        }
+        return stars;
+    }
+
+    public static Pair<Integer, Integer> getDisplayRange(
+            StatPossibleValues possibleValues, boolean showBestValueLastAlways) {
+        StatType statType = possibleValues.statType();
+        RangedValue valueRange = possibleValues.range();
+        boolean isGood = valueRange.low() > 0;
+        Pair<Integer, Integer> displayRange;
+        int first;
+        int last;
+        if (showBestValueLastAlways || isGood) {
+            first = valueRange.low();
+            last = valueRange.high();
+        } else {
+            // Emulate Wynncraft behavior by showing the value closest to zero first
+            first = valueRange.high();
+            last = valueRange.low();
+        }
+        // We store "inverted" stats (spell costs) as positive numbers internally,
+        // but need to display them as negative numbers
+        if (statType.showAsInverted()) {
+            first = -first;
+            last = -last;
+        }
+        displayRange = Pair.of(first, last);
+        return displayRange;
+    }
+
+    public static float getPercentage(StatActualValue actualValue, StatPossibleValues possibleValues) {
+        int min = possibleValues.range().low();
+        int max = possibleValues.range().high();
+
+        float percentage = MathUtils.inverseLerp(min, max, actualValue.value()) * 100;
+        return percentage;
+    }
+
+    public static double getPerfectChance(StatPossibleValues possibleValues) {
+        // FIXME: This is the chance of getting a *** (3 star) roll, not the chance of
+        // getting the maximum possible value. But for now, keep old behavior.
+        double perfectChance = 1 / (possibleValues.baseValue() > 0 ? 101d : 61d) * 100;
+        return perfectChance;
+    }
+
+    public static double getDecreaseChance(StatActualValue actualValue, StatPossibleValues possibleValues) {
+        assert !possibleValues.range().isFixed();
```

Comment:
just to confirm, this won't have any effect in a prod environment, right?

Reply:
No, unless the user for some reason adds `-ea` (enable asserts) to their command line.